### DISCUSSION TITLE
checking for detritus in the temp directory ... NOTE

### DISCUSSION THREAD

**Discussion by aitap** [r-devel/cran-cookbook#59] [Community Contributions]
Related to [Writing Files and Directories to the Home Filespace](https://contributor.r-project.org/cran-cookbook/code_issues.html#writing-files-and-directories-to-the-home-filespace).

**Problem**: your package got a `NOTE` complaining about the "detritus in the temp directory".

**Cause**: some of the code that runs as a result of `R CMD check` creates files in the temporary directory specified by the environment variable `TMPDIR` and doesn't remove them afterwards. This may include your own tests creating files under tempdir(), child processes, or Python code launched using `reticulate`.

**Solution**: if you use `tempfile()`/`tempdir()` as the default destination for functions called in your examples or tests, `unlink()` the resulting files as part of the test. In `\examples{}`, clean-up actions can be wrapped in `\dontshow{}` for aesthetic reasons. If you launch a browser, Calibre, or Python code, and it creates temporary files, find out what they are and remove them once the launched code is done with them. For example, R packages that use the `tensorflow` Python package [find and remove temporary files created by its auto-gradient feature](https://github.com/cran/vetiver/blob/35b24768cf0e84fab96610e001bba377dc777953/tests/testthat/setup.R#L13).

**Comment by hturner** [comment]
This one looks fine to me.

Probably worth mentioning `withr::local_tempfile()`/ `withr::local_tempdir()` that are "self-cleaning". These particularly make sense for {testthat} tests, as {testthat} imports {withr}, but `withr` is a low-dependency package that can be used independently. See https://r-pkgs.org/testing-design.html#sec-tests-files-where-write.

**Comment by jasdumas** [comment]
@aitap Thanks for this recipe! Please go ahead and add this as a PR. Here is the template guide to re-format your contribution: https://github.com/r-devel/cran-cookbook/wiki/Contributors-Guide#recipe-template-

**Comment by aitap** [comment]
Resolved in #69.
